### ISSUE TITLE
Support EVM specific functions for read-queries and write-queries

### ISSUE THREAD

**Issue by sanderpick** [tablelandnetwork/go-tableland#352]
There's been a few cases come up where it would be nice to grab some available state from transactions:

- `TXN_HASH()`: The validator would replace this text with the has of the txn that delivered the event (only available to write queries).
- `BLOCK_NUM()`: The validator would replace this text with the number of the block that delivered the event (only available to write queries).
- `BLOCK_NUM(<chain_id>)`: The validator would replace this text with the number of the _last seen_ block (only available to read queries). For example, in Rigs we want to show how long a rig has been _in-flight_ at a metadata attribute. We have `start_time` and `end_time` columns on a flight sessions table that is contains block numbers. Each flight session results in a row in this table with a start time. If `end_time` is `null`, the rig is in-flight. So, to calculate total flight time including an active session, we'd use this method:

```
SELECT SUM(COALESCE(end_time, BLOCK_NUM(<chain_id>))) - SUM(start_time) 
FROM pilot_sessions_80001_3515 WHERE rig_id=11
```

`BLOCK_TIME` could be added in the same way as `BLOCK_NUM`, with a read and write flavor.

**Comment by jsign** [comment]
@sanderpick ,  we're planning to support the described write-query functions `TXN_HASH()` and `BLOCK_NUM()`, and read-query functions `BLOCK_NUM(<chain_id>)` as described by OP.

We're pushing back on `BLOCK_TIME` not because it is hard to implement but because of a less obvious downside that I'll explain below.

Today in our execution pipeline, on each block, we get a list of transactions with events that we pipe into our execution pipeline. If we remember carefully, today, the block timestamp isn't required for executing a block. If we introduce this requirement (e.g: to resolve potential `BLOCK_TIME()` calls), we're forced to fetch the block timestamp.

What probably isn't apparent is that we can't leverage any existing calls in the execution pipeline that already return this event. We receive the block with the transactions with events to execute in a single Ethereum API call that **do not** contain this information. This means that for every received block, we have to do a new API call to fetch the complete block information that contains the timestamp.

If we do this, we'll slow down our execution pipeline by orders of magnitude. Today our execution pipeline asks in a single Ethereum API call for a _range_ of blocks that contain transactions with events. So today we're able to process, for example, 100 blocks, only paying the cost of doing a single Ethereum API call.

If we now introduce the requirement that every received block also do an extra call, we'll be doing 100 extra Ethereum API calls, one per received block.

So, if for 100 blocks, we were doing one Ethereum API call, we'll jump to doing 101 API calls.
This has two downsides:
1) We'll be severely increasing the Ethereum API cost of the used provider.
2) More importantly, we put a bound on how fast we can process blocks since an external Ethereum API call to get block information will have a lower bound of, say, 50ms in the best case.

Realizing this wasn't surprising because I faced the same challenge when including the _optional_ analytics flag in the validator (months ago), which fetches extra information to build analytics data. When analytics are enabled, we have to do this extra API call to get the timestamp of each block (for analytics reasons). 

In the analytics case, I made this extra information collecting async in the background not to slow down the execution pipeline for the reasons described earlier. This is fine since we're OK with having lagged analytics information for a while. But in the case of having the block timestamp needed for **execution**, we can't apply this same idea since having that information is blocking the execution of the transaction.

cc @carsonfarmer  FYI.
